clients/spark/core/src/main/scala/io/treeverse/clients/URLResolver.scala

```diff
@@ -0,0 +1,11 @@
+package io.treeverse.clients
+
+import java.net.URI
+import org.apache.hadoop.fs.Path
+
+object URLResolver extends Serializable{
+  def resolveURL(uri: URI, extraPath: String): Path = {
+    val newPath: String = uri.getPath + '/' + extraPath
```

Comment:
I don't understand this function.  It seems to me it either does what `uri.resolve(extraPath)` does when extraPath is relative, or some _else_ (strange but in a different manner) when `extraPath` is absolute.  And it does something *really* odd when `extraPath` starts with a protocol.

Reply:
This function was just moved to another file (see explanation in previous comment):

> it either does what uri.resolve(extraPath) does when extraPath is relative

Nope, for `uri=s3://bucket/with/prefix`, `extraPath=/foo/bar`, `uri.resolve(extraPath)` will return (surprise surprise) `s3://bucket/foo/bar`.

> when extraPath is absolute

Didn't think about this one. I'll document against such usage

Reply:
> This function was just moved to another file (see explanation in previous comment):
> 
> > it either does what uri.resolve(extraPath) does when extraPath is relative
> 
> Nope, for `uri=s3://bucket/with/prefix`, `extraPath=/foo/bar`, `uri.resolve(extraPath)` will return (surprise surprise) `s3://bucket/foo/bar`.

Sorry, bad usage of the word "relative".  Should have said "URI with a relative path", which is a special case of a relative URL.  `extraPath` above is a relative URI with an absolute path.

To be clear(-er): I claim that a URI with a trailing slash resolves URIs with relative paths as resolveURL does.

The function above returns `s3://bucket/fwith/prefix//foo/bar`; note the double slashes / "empty" "directory".

> 
> > when extraPath is absolute
> 
> Didn't think about this one. I'll document against such usage

To summarize: This function does _the same thing_ as `URI.resolve` in the common case, except that it *forbids* a slash at the end of `uri`, whereas using `URI` like this *requires* it.  In all the uncommon cases (`extraPath` starts with a slash, uri ends/doesn't end in a slash, etc.), both do something "unexpected" in the sense that the code probably doesn't work.  I would argue that `URI.resolve` is _well-defined_ and _standard_.  We might be able to achieve parity on definitions, but not on standardization.  So `URI.resolve` wins to my understanding.